common/lib/xmodule/xmodule/modulestore/mixed.py

```diff
@@ -157,35 +157,39 @@ def __init__(
                     self.mappings[course_key] = store
             self.modulestores.append(store)
 
-    def _clean_course_id_for_mapping(self, course_id):
+    def _clean_locator_for_mapping(self, locator):
         """
-        In order for mapping to work, the course_id must be minimal--no version, no branch--
+        In order for mapping to work, the locator must be minimal--no version, no branch--
         as we never store one version or one branch in one ms and another in another ms.
 
-        :param course_id: the CourseKey
+        :param locator: the CourseKey
         """
-        if hasattr(course_id, 'version_agnostic'):
-            course_id = course_id.version_agnostic()
-        if hasattr(course_id, 'branch'):
-            course_id = course_id.replace(branch=None)
-        return course_id
+        if hasattr(locator, 'version_agnostic'):
+            locator = locator.version_agnostic()
+        if hasattr(locator, 'branch'):
+            locator = locator.replace(branch=None)
+        return locator
 
-    def _get_modulestore_for_courseid(self, course_id=None):
+    def _get_modulestore_for_courselike(self, locator=None):
         """
-        For a given course_id, look in the mapping table and see if it has been pinned
+        For a given locator, look in the mapping table and see if it has been pinned
         to a particular modulestore
 
-        If course_id is None, returns the first (ordered) store as the default
+        If locator is None, returns the first (ordered) store as the default
         """
-        if course_id is not None:
-            course_id = self._clean_course_id_for_mapping(course_id)
-            mapping = self.mappings.get(course_id, None)
+        if locator is not None:
+            locator = self._clean_locator_for_mapping(locator)
+            mapping = self.mappings.get(locator, None)
             if mapping is not None:
                 return mapping
             else:
+                if isinstance(locator, LibraryLocator):
+                    has_course = lambda store: hasattr(store, 'has_library') and store.has_library(locator)
+                else:
+                    has_course = lambda store: store.has_course(locator)
                 for store in self.modulestores:
-                    if store.has_course(course_id):
-                        self.mappings[course_id] = store
+                    if has_course(store):
```

Comment:
@bradenmacdonald nit: something like `has_courselike(store, locator)` would be more self-explaining here, so you might want changing the name and signature of this lambda.